processing/stenosis_connector.go

```diff
@@ -0,0 +1,185 @@
+package processing
+
+// DCSO FEVER
+// Copyright (c) 2020, DCSO GmbH
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/DCSO/fever/stenosis/api"
+	"github.com/DCSO/fever/stenosis/task"
+	"github.com/DCSO/fever/types"
+	"github.com/golang/protobuf/ptypes"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	cache "github.com/patrickmn/go-cache"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	defaultStenosisTimeBracket = 1 * time.Minute
+	maxErrorCount              = 10
+)
+
+// StenosisConnector is a handler that caches alerts and waits for the
+// associated flow to finish, then annotates all alerts with flow IDs and
+// performs queries against a specified Stenosis server. Alerts will then
+// be annotated with returned tokens and forwarded.
+type StenosisConnector struct {
+	Endpoint       string
+	Client         api.StenosisServiceQueryClient
+	TimeBracket    time.Duration
+	Timeout        time.Duration
+	ErrorCount     uint64
+	FlowNotifyChan chan types.Entry
+	Cache          *cache.Cache
+}
+
+// MakeStenosisConnector returns a new StenosisConnector for the
+// given parameters.
+func MakeStenosisConnector(endpoint string, timeout, timeBracket time.Duration,
+	notifyChan chan types.Entry, forwardChan chan []byte,
+	tlsConfig *tls.Config) (*StenosisConnector, error) {
+	sConn := &StenosisConnector{
+		Endpoint:       endpoint,
+		FlowNotifyChan: notifyChan,
+		TimeBracket: func() time.Duration {
+			if timeBracket != 0 {
+				return timeBracket
+			}
+			return defaultStenosisTimeBracket
+		}(),
+		Timeout: timeout,
+		//TODO: make configurable
+		Cache: cache.New(30*time.Minute, 30*time.Second),
+	}
+	dialOpts := make([]grpc.DialOption, 0, 1)
+	if tlsConfig != nil {
+		dialOpts = append(dialOpts, grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
+	} else {
+		dialOpts = append(dialOpts, grpc.WithInsecure())
+	}
+	conn, err := grpc.Dial(endpoint, dialOpts...)
+	if err != nil {
+		return nil, err
+	}
+	sConn.Client = api.NewStenosisServiceQueryClient(conn)
+
+	go func() {
+		defer conn.Close()
+
+		for flow := range sConn.FlowNotifyChan {
+			var myAlerts []types.Entry
+			aval, exist := sConn.Cache.Get(flow.FlowID)
+			if exist {
+				log.Debugf("flow with existing alert finished: %v", flow)
+				myAlerts = aval.([]types.Entry)
+				outParsed, err := sConn.submit(&flow)
+				if err != nil {
+					log.Error(err)
+					for _, a := range myAlerts {
+						forwardChan <- []byte(a.JSONLine)
+					}
+				} else {
+					if forwardChan != nil {
+						for _, a := range myAlerts {
+							var ev types.EveOutEvent
+							// annotate alerts with flows and forward
+							err := json.Unmarshal([]byte(a.JSONLine), &ev)
+							if err != nil {
+								log.Error(err)
```

Comment:
True. Again, just for defensive programming, added a `continue`. We know that the JSON can be (un)marshaled because the rest of the struct was populated from it.